src/tests/JIT/opt/Regressions/Regression2.cs

```diff
@@ -0,0 +1,83 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Runtime.CompilerServices;
+
+public class Program
+{
+    [MethodImpl(MethodImplOptions.NoInlining)]
+    static void Add_SmallType_Correctness()
+    {
+        for (int i = 0; i < ushort.MaxValue + 1; i++)
+        {
+            for (int j = 0; j < ushort.MaxValue + 1; j++)
+            {
+                if ((byte)(i + j) != (byte)((byte)i + (byte)j))
+                {
+                    throw new Exception();
+                }
+            }
+        }
+    }
+
+    [MethodImpl(MethodImplOptions.NoInlining)]
+    static void Sub_SmallType_Correctness()
+    {
+        for (int i = 0; i < ushort.MaxValue + 1; i++)
+        {
+            for (int j = 0; j < ushort.MaxValue + 1; j++)
+            {
+                if ((byte)(i - j) != (byte)((byte)i - (byte)j))
+                {
+                    throw new Exception();
+                }
+            }
+        }
+    }
+
+    [MethodImpl(MethodImplOptions.NoInlining)]
+    static void Mul_SmallType_Correctness()
+    {
+        for (int i = 0; i < ushort.MaxValue + 1; i++)
+        {
+            for (int j = 0; j < ushort.MaxValue + 1; j++)
+            {
+                if ((byte)(i * j) != (byte)((byte)i * (byte)j))
+                {
+                    throw new Exception();
+                }
+            }
+        }
+    }
+
+    [MethodImpl(MethodImplOptions.NoInlining)]
+    static void Div_SmallType_Correctness()
+    {
+        for (int i = 0; i < ushort.MaxValue + 1; i++)
+        {
+            for (int j = 1; j < ushort.MaxValue + 1; j++)
+            {
+                if ((byte)(i / j) != (byte)((byte)i / (byte)j))
+                {
+                    throw new Exception();
+                }
+            }
+        }
+    }
+
+    static int Main()
+    {
+        Add_SmallType_Correctness();
+        Sub_SmallType_Correctness();
+        Mul_SmallType_Correctness();
+
+        try
+        {
+            Div_SmallType_Correctness();
+        }
+        catch(DivideByZeroException) {}
```

Comment:
This is going to stop executing at `i==0, j==256` and skip all of the other cases.

Reply:
Yea it looks to be that way because when `j` is 256, then when its cast, the value is `0` and we get the divide by zero. I'll get rid of the first `for` loop iteration.

Reply:
That doesn't seem to be what the intention was here.  These four were attempting to test all combinations of ushorts.  To do that, you need to change the Div function to do something like

```
try
{
    if ((byte)(i / j) != (byte)((byte)i / (byte)j))
    {
        throw new Exception();
    }
}
catch (DivideByZeroException)
{
    if ((byte)j != 0)
    {
        throw new Exception();
    }
}
```
In this case, you can change the inner loop back to start with zero.

However, this seems pretty excessive trying every pair of ushorts (4 billion iterations per function).  How about choosing a set of interesting values and iterating over them?

```
var interesting = new ushort[] { 0, 1, 2, 13, 255, 256, 511, 512, whatever };
foreach (int i in interesting) { foreach (int j in interesting) { /* operation */ } }
```

Reply:
It does seem excessive. I think maybe the most interesting test case for this is basically what you mentioned before with `i` and `j`:
```if ((byte)(0 / 256) != (byte)((byte)0 / (byte)256))```

The test is ensuring that `(byte)256` the cast will indeed happen and not get removed.

Reply:
I'll just simplify it down to that one case as it is really what we are testing for.